### PR TITLE
fix(app): update ODD cancel run modal run dismissal and routing

### PR DESCRIPTION
**Pull request by brenthagen**
# Overview

run dismissal from the cancel run modal is giving a 409 ("Current run is not idle or stopped.") because the modal calls dismiss run before the robot is in a stopped status. this results in erratic routing behavior after the cancel run modal routes to the protocol details page, because the stopped run is still current.

this PR adds a useEffect on the cancel run modal checking for a stopped run status before dismissing, and changes some run routing conditions in the top level redirects.

additionally, going back from the protocol details page should navigate to its parent node, the protocols list page.

closes RQA-1174

# Test Plan

 - manually verified canceling a run from the protocol setup page dismisses the current run and navigates to the protocol details page
 - manually verified protocol details back button goes to protocols list page

# Changelog

 - Cancel run modal dismisses the current run on stopped run status
 - Protocol details back button navigates to protocols list page

# Review requests

 - confirm that canceling a run from the protocol setup page results in correct routing to the protocol details page 
 - confirm that canceling a run during a run is unaffected

# Risk assessment

low

### CODE REVIEW

**Review by sfoster1 — APPROVED** [review]
Looks good!

**Review by koji — APPROVED** [review]
lgtm

File: app/src/App/OnDeviceDisplayApp.tsx [hunk 1]
```diff
@@ -292,11 +292,7 @@ export const OnDeviceDisplayApp = (): JSX.Element => {
 }
 
 function TopLevelRedirects(): JSX.Element | null {
-  const runRouteMatch = useRouteMatch({ path: '/runs/:runId' })
   const currentRunRoute = useCurrentRunRoute()
-
-  if (runRouteMatch != null && currentRunRoute == null)
```
Comment (brenthagen) [inline app/src/App/OnDeviceDisplayApp.tsx:302]: @b-cooper is this dashboard redirect safe to remove? we need to hang around on the cancel run modal until the run status is stopped to dismiss the run (see changes to `useCurrentRunRoute` and the `ConfirmCancelRunModal`)

**Comment by brenthagen** [comment]
@koji @sfoster1 PR updated because we were getting a 409 on run dismissal from the cancel run modal. tested from the run setup page but should probably also be tested when cancelling during a run.

File: app/src/organisms/OnDeviceDisplay/RunningProtocol/ConfirmCancelRunModal.tsx [hunk 2]
```diff
@@ -56,23 +59,26 @@ export function ConfirmCancelRunModal({
...
+      dismissCurrentRun(runId)
+      if (!isActiveRun) {
+        if (protocolId != null) {
+          history.push(`/protocols/${protocolId}`)
+        } else {
+          history.push(`/protocols`)
+        }
+      }
+    }
+  }, [runStatus])
```
Comment (brenthagen) [inline app/src/organisms/OnDeviceDisplay/RunningProtocol/ConfirmCancelRunModal.tsx:80]: similar pattern to what we use in desktop `ProtocolRunHeader`

**Review by shlokamin — APPROVED** [review]
tested on my dev kit, works as expected

**Comment by brenthagen** [comment]
merging on hold pending robot smoke testing

**Pull request merged**